captum/attr/_utils/gradient.py

```diff
@@ -175,32 +168,106 @@ def forward_hook(module, inp, out=None):
             # TODO we need to think what will be the best way of storing eval
             # tensors per device for each input per example. This implementation
             # doesn't support a tuple of inputs
-            saved_layer[eval_tsr.device] = eval_tsr.clone()
+            if forward_hook_with_return:
+                saved_layer[eval_tsr.device] = eval_tsr
+                return eval_tsr.clone()
+            else:
+                saved_layer[eval_tsr.device] = eval_tsr.clone()
 
     if attribute_to_layer_input:
         hook = layer.register_forward_pre_hook(forward_hook)
     else:
         hook = layer.register_forward_hook(forward_hook)
-    _run_forward(forward_fn, inputs, additional_forward_args=additional_forward_args)
+    output = _run_forward(
+        forward_fn,
+        inputs,
+        target=target_ind,
+        additional_forward_args=additional_forward_args,
+    )
     hook.remove()
+
     if len(saved_layer) == 0:
         raise AssertionError("Forward hook did not obtain any outputs for given layer")
 
+    if forward_hook_with_return:
+        return saved_layer, output
+    return saved_layer
+
+
+def _gather_distributed_tnsrs(saved_layer, device_ids=None, key_list=None):
+    r"""
+    A helper function to concatenate intermediate layer results stored on
+    different devices in `saved_layer`. `saved_layer` is a dictionary that
+    that contains `device_id` as a key and intermediate layer results (either
+    the input or the output of the layer) stored on the device corresponding to
+    the key.
+    `key_list` is a list of devices in appropriate ordering for concatenation
+    and in case it is not provided, it is identified based on device ids.
```

Comment:
Nit: `and in case it is not provided, it is identified based on device ids.` ->  `if not provided, keys are sorted based on device ids.`